### PR TITLE
Clone and transfer volumes between Docker hosts

### PR DESCRIPTION
**Pull request by felipecruz91**
This PR adds the following two new functionalities:

- **Clone** locally an existing volume.
image

- **Transfer** a volume between Docker hosts.

image

### CODE REVIEW

File: ui/src/App.tsx [hunk 1]
```diff
@@ -102,6 +110,17 @@ export function App() {
           onClick={handleNavigate(params.row)}
           disabled={actionInProgress}
         />,
+        <GridActionsCellItem
```
Comment (benja-M-1) [inline ui/src/App.tsx:113]: the list of actions is growing quite a lot, don't you think a [3 dots menu](https://mui.com/material-ui/react-menu/) would be better?

File: ui/src/components/CloneDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+import { MyContext } from "../index";
+
+const client = createDockerDesktopClient();
+
+function useDockerDesktopClient() {
+  return client;
+}
+
+export default function CloneDialog({ ...props }) {
+  console.log("CloneDialog component rendered.");
```
Comment (benja-M-1) [inline ui/src/components/CloneDialog.tsx:26]: nit: remove

File: ui/src/components/CloneDialog.tsx [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  );
+  const [actionInProgress, setActionInProgress] =
+    React.useState<boolean>(false);
+
+  const cloneVolume = async () => {
+    setActionInProgress(true);
+
+    try {
+      // TODO: check if destination volume already exists
+      console.log(`Creating destination volume ${volumeName}...`);
```
Comment (benja-M-1) [inline ui/src/components/CloneDialog.tsx:41]: not: remove

File: ui/src/components/CloneDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        "ash",
+        "-c",
+        '"cd /from ; cp -av . /to"',
+      ]);
+      if (cloneOutput.stderr !== "") {
+        ddClient.desktopUI.toast.error(cloneOutput.stderr);
+        return;
+      }
+
+      console.log(cloneOutput);
```
Comment (benja-M-1) [inline ui/src/components/CloneDialog.tsx:69]: nit: remove

File: ui/src/components/CloneDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+      ]);
+      if (createVolumeOutput.stderr !== "") {
+        ddClient.desktopUI.toast.error(createVolumeOutput.stderr);
+        return;
+      }
+      console.log(createVolumeOutput);
+
+      console.log(
+        `Copying data from source volume ${context.store.volumeName} to destination volume ${volumeName}...`
+      );
```
Comment (benja-M-1) [inline ui/src/components/CloneDialog.tsx:54]: nit: remove

File: ui/src/components/TransferDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+import { MyContext } from "../index";
+
+const client = createDockerDesktopClient();
+
+function useDockerDesktopClient() {
+  return client;
+}
+
+export default function TransferDialog({ ...props }) {
+  console.log("TransferDialog component rendered.");
```
Comment (benja-M-1) [inline ui/src/components/TransferDialog.tsx:27]: nit: remove

File: ui/src/components/TransferDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+  useEffect(() => {
+    let active = true;
+
+    if (!autocompleteLoading) {
+      return undefined;
+    }
+
+    (async () => {
+      const volumes = await listVolumesForDockerHost();
+      console.log(volumes);
```
Comment (benja-M-1) [inline ui/src/components/TransferDialog.tsx:49]: nit: remove

File: ui/src/components/TransferDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+  }, [autocompleteLoading]);
+
+  useEffect(() => {
+    if (!autocompleteOpen) {
+      setOptions([]);
+    }
+  }, [autocompleteOpen]);
+
+  const listVolumesForDockerHost = async () => {
+    console.log("Listing volumes for Docker host");
```
Comment (benja-M-1) [inline ui/src/components/TransferDialog.tsx:68]: nit: remove

File: ui/src/components/TransferDialog.tsx [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+    }
+  };
+
+  const transferVolume = async () => {
+    setActionInProgress(true);
+
+    try {
+      console.log(
+        `Transferring data from source volume ${context.store.volumeName} to destination volume ${volumeName} in host ${destHost}...`
+      );
```
Comment (benja-M-1) [inline ui/src/components/TransferDialog.tsx:97]: nit: remove
If it is useful, can't it be a message printed to the user?

File: ui/src/components/TransferDialog.tsx [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+        "ash",
+        "-c",
+        `"cd /from ; tar -czf - . \" | ssh ${destHost} \"docker run --rm -i -v \"${volumeName}\":/to alpine ash -c 'cd /to ; tar -xpvzf - '"`,
+      ]);
+      if (transferredOutput.stderr !== "") {
+        ddClient.desktopUI.toast.error(transferredOutput.stderr);
+        return;
+      }
+
+      console.log(transferredOutput);
```
Comment (benja-M-1) [inline ui/src/components/TransferDialog.tsx:118]: nit: remove

File: ui/src/components/CloneDialog.tsx (hunk 5 above)
Comment (benja-M-1) [reply ui/src/components/CloneDialog.tsx:54]: if it is useful, maybe you should add this message in the UI? See https://mui.com/material-ui/react-snackbar/#customized-snackbars

**Review by benja-M-1 — COMMENTED** [review]
Seems good, I only read the code and didn't notice anything wrong

File: ui/src/App.tsx (hunk 1 above)
Comment (felipecruz91) [reply ui/src/App.tsx:114]: I've moved some actions to the 3 dots menu, thanks!

**Pull request merged**